[Security Solution] add error handling for buildEsQuery calls

kapral18 opened the pull request:
This PR addresses issue #155419 by adding missing error handling to `buildEsQuery()` calls within investigation/explore domains.

This PR underwent numerous iterations and remained in draft status for an extended period before reaching it's final shape. Initially, many changes were made to the `useInvalidFilterQuery()` hook, which manages the notification toasts in various locations under an assumption that some error toasts are not invoked when calling `buildEsQuery()`. During iteration the hook was corrected, and tests were written to better comprehend its behavior.

After eventual realization that error toasts are invoked in most of the cases and subsequent cancellation of most of the work related to `useInvalidFilterQuery()` calls, it was decided to retain the tests and refactoring that occurred during the process as part of this PR anyways.

Scope:
The only file affected after numerous scope cuts is this `x-pack/plugins/security_solution/public/detections/hooks/trigger_actions_alert_table/use_alert_actions.tsx`

kapral18 committed:
[SECURITY SOLUTION] improve/fix test provider mock

kapral18 committed:
[SECURITY SOLUTION] fix incorrect prop name for filterQuery

kapral18 committed:
[SECURITY SOLUTION] refactor use_invalid_filter_query and add tests

kapral18 committed:
[SECURITY SOLUTION] add buildEsQuery to try block

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

elasticmachine commented:
Pinging @elastic/security-threat-hunting-explore (Team:Threat Hunting:Explore)

elasticmachine commented:
Pinging @elastic/security-threat-hunting-investigations (Team:Threat Hunting:Investigations)

semd commented on x-pack/plugins/security_solution/public/common/hooks/use_invalid_filter_query.test.tsx:
```diff
@@ -0,0 +1,298 @@
+  it('invokes error toast with error title and error instance without original stack', () => {
+    const store = getStore();
+    const props = getProps();
+    const InvalidFilterComponent = () => {
+      useInvalidFilterQuery(props);
+      return null;
+    };
+    render(
+      <TestProviders store={store} startServices={kibanaMock}>
+        <InvalidFilterComponent />
+      </TestProviders>
+    );
```
Instead of creating a proxy function component in every test case to render the hook indirectly, we can also use _renderHook_ from '@testing-library/react-hooks' to render the hook directly:

```suggestion
    renderHook(useInvalidFilterQuery, { initialProps: props, wrapper: getWrapper(store) });
```

We can also extract the wrapper creation into:

```
const getWrapper = (store: Store): React.FC => {
  // eslint-disable-next-line react/display-name
  return ({ children }) => (
    <TestProviders store={store} startServices={kibanaMock}>
      {children}
    </TestProviders>
  );
};
```

That's how we usually do it. 
What do you think?

kapral18 replied:
Absolutely agree, thanks.

semd commented on x-pack/plugins/security_solution/public/common/lib/kibana/kibana_react.mock.ts:
```diff
@@ -266,10 +266,16 @@ export const createKibanaContextProviderMock = () => {
   // eslint-disable-next-line react/display-name
-  return ({ children }: { children: React.ReactNode }) =>
+  return ({
+    children,
+    startServices: startServicesMock,
+  }: {
+    children: React.ReactNode;
+    startServices?: StartServices;
+  }) =>
     React.createElement(
       KibanaContextProvider,
-      { services },
+      { services: startServicesMock || services },
```
❤️

kapral18 committed:
[SECURITY SOLUTION] revert test writing style and add new tests

- revert test writing style from bdd back to regular
- refactor to use renderHook instead of render
- add new tests that test behavior across rerenders

semd reviewed (approved):
LGTM! Thanks Karen 🚀

semd commented:
@elasticmachine merge upstream

kapral18 commented:
@elasticmachine merge upstream

logeekal commented on x-pack/plugins/security_solution/public/common/hooks/use_invalid_filter_query.tsx:
```diff
@@ -13,24 +13,35 @@ import { appActions } from '../store/app';
 import { useAppToasts } from './use_app_toasts';
 import { useDeepEqualSelector } from './use_selector';
 
+export const genHash = (message: string) =>
```
`nit` : I know you just extraacted this code to make a reusable function, Could you please add the purpose of this hash. I am not sure I 100% understand what this bitwise operation is doing.

Thanks.

kapral18 replied:
the purpose of most hash string functions is to convert a string of variable size to a string of a fixed size for further comparison. Is this what you mean when asking about the purpose? Or do you mean like why do we need the hash here in the first place? If the latter then I don't know it's a good question to the author. Because from my observations you probably can go by without a hash and just compare the original message strings altogether.

logeekal commented on x-pack/plugins/security_solution/public/common/hooks/use_invalid_filter_query.test.tsx:
```diff
@@ -0,0 +1,286 @@
+  // when you have different errors with the same id,
+  // but this test shows that it will currently show multiple toasts
+  // so we need to double check if this is the intended behavior and fix otherwise
+  it('invokes toast for each error, when called multiple times with same id and different errors, during a single render', async () => {
```
I have not been able to reproduce it.. For example, in the demo below, I see multiple errors to the query but I see only one toast. Do you think it might be worth checking why that is the case.

kapral18 replied:
Hey Jatin, what do you mean by "demo below"?

logeekal replied:
@kapral18 , Sorry I missed attaching video in my last comment. Just added it.

logeekal commented:
>which manages the notification toasts in various locations under an assumption that some error toasts are not invoked when calling buildEsQuery()

Thanks @kapral18  for making these changes. I did not completely understand above statemen for you PR description. Could you please expand on what `error toasts` are not invoked when calling `buildEsQuery` ?

Thanks.

logeekal commented on x-pack/plugins/security_solution/public/common/hooks/use_invalid_filter_query.test.tsx:
```diff
@@ -0,0 +1,286 @@
+    jest.clearAllMocks();
+  });
+
+  it('invokes error toast with error title and error instance without original stack', () => {
```
Thanks for adding plethora of these tests. This is awesome. 🚀

logeekal commented on x-pack/plugins/security_solution/public/overview/components/events_by_dataset/index.tsx:
```diff
@@ -138,16 +138,16 @@ const EventsByDatasetComponent: React.FC<Props> = ({
   );
 
   const [filterQuery, kqlError] = useMemo(() => {
```
`nit` : Here `kqlError` will always be undefined. Do you think it should be removed?

kapral18 replied:
2 things:

1. I didn't touch this code, and I thought I already touched too much out of scope changes in other places so cleaning up even more was probably unnecessary to not further delay this already super delayed ticket :)

2. Why do you think this will return `undefined`, can you elaborate?

logeekal replied:
Sorry.. This is my mistake. Please ignore this comment.

logeekal reviewed (commented):
Thank @kapral18 for making these changes and tests. I have posted my few comments/questions below. None of them are blocker but I wanted to understand this PR more before I approve it.

Additionaly, does this PR also intends to convert cases like below where error handling when calling `buildEsQuery` is absent.

https://github.com/elastic/kibana/blob/eadc173ba23e44ed30e3dcfdf8b3b01c683b9358/x-pack/plugins/security_solution/public/flyout/document_details/shared/hooks/use_fetch_prevalence.ts#L131-L134

kapral18 commented:
> > which manages the notification toasts in various locations under an assumption that some error toasts are not invoked when calling buildEsQuery()
> 
> Thanks @kapral18 for making these changes. I did not completely understand above statemen for you PR description. Could you please expand on what `error toasts` are not invoked when calling `buildEsQuery` ?
> 
> Thanks.

good question @logeekal 

Initially when I started working on the ticket I reached out to the stakeholders and was told to make sure on top of adding try catches, to also double check that error toasts are called. Basically to make sure that we don't have an instance of try+catch without showing the respective toast. But after testing on the UI to the best of my current ability :) I couldn't reproduce the assumption that we might have a missing toast, all toasts seem to be in place already. And this `useInvalidFilterQuery` was the main utility that was properly operating with those toasts taking into account hashing messages and storing them in redux store to make sure we don't end up duplicating error toasts all over the place when called from multiple places.

Let me know if this explanation makes sense.

logeekal commented:
>we decided to not try catch buildEsQuery calls that are using static/predetermined input because we noticed that error atm only happens from an invalid user formulated query. (cc: @semd )

This makes sense. Thanks.

logeekal reviewed (approved):
Changes + tests looks great 🚀 . I am still not 100% sure of the scope that this PR upholds. But I trust that @semd  has already reviewed it from that perspective.

Thanks for shootting down these bugs.

kapral18 commented:
> Changes + tests looks great 🚀 . I am still not 100% sure of the scope that this PR upholds. But I trust that @semd has already reviewed it from that perspective.
> 
> Thanks for shootting down these bugs.

@logeekal scope is pretty small, because we repeatedly cut it. I started with an intention to cover the entire security_solution, and then we decide to shrink it to investigation + explore only, and then we shrink it even more to exclude requests with no queries but filters because filters are controlled so very little chance of erroring, and the only thing remained was dynamic queries from user which is essentially everything that user types in kql bar, and for that criterion the only file that matched was this `x-pack/plugins/security_solution/public/detections/hooks/trigger_actions_alert_table/use_alert_actions.tsx` that I just put into an existing try catch that shows the toast and that's it. All other changes are a side effect of figuring things out and realizing those changes are not needed, but the byproduct of the effort in a form of fixing names, cleanup and adding tests I decided to leave as is.

logeekal commented:
>scope is pretty small, because we repeatedly cut it .....

Thanks for explanation. It clears things up for me. 

Consider updating the PR summary with clear final scope of the PR.
